cpp/src/arrow/array/array_dict.cc

```diff
@@ -44,6 +44,29 @@ namespace arrow {
 using internal::checked_cast;
 using internal::CopyBitmap;
 
+struct MaxIndexAccessor {
```

Comment:
The DictionaryUnifier is typed to the value type of the dictionary but not the index type so I needed to create this utility for accessing the max possible value for a given index type.  I wasn't really sure where to put this.  There could also be a general purpose visitor for finding the min/max of all numeric types.  Also, do we have this capability anywhere else in the code base I could just leverage?  This is needed by DictionaryUnifierImpl::GetResultWithIndexType

Reply:
Even better, it turns out `util/int_util.h` already had what I needed.  I switched to using `internal::IntegersCanFit`.